common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java

```diff
@@ -45,18 +86,24 @@ public int read(ByteBuffer dst) throws IOException {
 
     if (data.readableBytes() == 0) {
       data.release();
+      data = null;
     }
 
     return totalRead;
   }
 
   @Override
   public void close() throws IOException {
+    closed = true;
+    if (data != null) {
+      data.release();
```

Comment:
This feels like the wrong place to do it. Shouldn't `TransportCipher.channelRead` do that instead after it decrypts the buffer? (Then you also don't need L46.)

Reply:
@vanzin again this class is public and so I think it is not guaranteed that read(...) will always consume the whole buffer.  The `ByteArrayReadableChannel` is wrapped in a `CryptoInputStream` and who knows if it always is guaranteed that it will consume all data (for example there is at least one case where you catch an `InternalError` which may cause that the previous feed data was not released (as it was not fully consumed).

So with all this in mind I think it is the safest to at least ensure we release everything once `close()` is called.